Extend the definition of sensitive-fields in the Database model to include any custom password type properties

jeff303 opened the pull request:
Change the type of custom-fields in database.clj to a set instead of vector

Change references to the custom-fields set to instead call the new function

Adding test to confirm the a custom password field is included

jeff303 commented:
This is for metabase/metaboat#31.

robdaemon commented on src/metabase/models/database.clj:
```diff
@@ -155,9 +155,32 @@
+  [database]
+  (if (empty? database)
+    (get-sensitive-fields nil) ;; preserve existing behavior by returning the default sensitive fields
+    (get-sensitive-fields (driver.u/database->driver database))))
```
Is this better suited as a `(defmulti)` that can be implemented in each driver? See something like: https://github.com/metabase/metabase/blob/master/src/metabase/driver.clj#L416-L432

camsaul replied:
I don't think this is something we'd want to use `defmulti` for -- this is just derived directly from `metabase.driver/connection-properties` which is already a multimethod. I don't see a situation where we'd have a different list of sensitive fields in `connection-properties` vs what comes back here

jeff303 replied:
Have a look at the refactor.

robdaemon commented on test/metabase/models/database_test.clj:
```diff
@@ -11,7 +11,8 @@
             [schema.core :as s]
-            [toucan.db :as db]))
+            [toucan.db :as db]
+            [metabase.driver :as driver]))
```
namespace sorting

jeff303 replied:
Fixed in refactor.

robdaemon reviewed (dismissed):
Overall, super good. This feature definitely needs this!

Just the one change to make it a multi-method like the other drivers. @camsaul? What's your thoughts on that?

jeff303 commented:
> Overall, super good. This feature definitely needs this!
> 
> Just the one change to make it a multi-method like the other drivers. @camsaul? What's your thoughts on that?

Are people able to define multimethod implementations for custom driver plugins?  The context here is metabase/metaboat#31, where the snippet shown is from a YAML file (I'm not familiar with this mechanism; is it similar to the non-core drivers in that there is Clojure code as well?)

camsaul commented:
Sorry for the late reply.

Yes, people are able to define multimethod implementations for custom drivers. The YAML manifest just tells Metabase what the driver is called, how to load it (and if it has any parent drivers/prereqs), and the connection details so we can power the UI without having to load 100MB of classes (the whole reason we have the driver plugin system in the first place is the Hive/Spark SQL JDBC driver includes a full copy of Hadoop and a bunch of other classes that we can't exclude without breaking stuff).

When you actually add a database of that type and sync/query it we load up all the namespaces automatically. 

:+1: on making stuff multimethods

jeff303 commented:
> Sorry for the late reply.
> 
> Yes, people are able to define multimethod implementations for custom drivers. The YAML manifest just tells Metabase what the driver is called, how to load it (and if it has any parent drivers/prereqs), and the connection details so we can power the UI without having to load 100MB of classes (the whole reason we have the driver plugin system in the first place is the Hive/Spark SQL JDBC driver includes a full copy of Hadoop and a bunch of other classes that we can't exclude without breaking stuff).
> 
> When you actually add a database of that type and sync/query it we load up all the namespaces automatically.
> 
> 👍 on making stuff multimethods

OK, this code has been updated such that there is a multimethod called `get-sensitive-fields` (and all related stuff is now moved to driver namespace, where it seems like it should go at this point).  The default implementation is a function that does what was outlined above (includes a set of default field names plus any other that have `:type :password`).  Backwards compatibility is kept in a fairly paranoid manner, so that in case there is a `nil` or `empty?` db spec, then it will still just return the default set that we were using before this patch.

camsaul commented:
> OK, this code has been updated such that there is a multimethod called `get-sensitive-fields`

See https://github.com/metabase/metabase/wiki/Metabase-Clojure-Style-Guide#pure-function-names-should-be-nouns or https://stuartsierra.com/2016/01/09/how-to-name-clojure-functions, it's not really idiomatic to put verbs like `get-` in front of function names like that

camsaul commented on src/metabase/driver.clj:
```diff
@@ -594,3 +594,31 @@
+  dispatch-on-initialized-driver
+  :hierarchy #'hierarchy)
+
+(def ^:const default-sensitive-fields
```
Even tho this _is_ a constant, I would advise against making it `^:const`. This just means it gets inlined into everywhere that uses it. So that might actually make performance worse if we end up with more code overall. It will make testing or REPL-based development with it harder since you can't `with-redefs` it, or if you change the value anything using it won't pick up the changes until you reload it

In general `^:const` makes most sense for simple constants like integers

camsaul replied:
I guess the old code had this marked `^:const`... I would remove that

jeff303 replied:
Removed in the refactor.

camsaul commented on src/metabase/driver.clj:
```diff
@@ -594,3 +594,31 @@
+    :access-token :refresh-token :service-account-json})
+
+(defn default-get-sensitive-fields
+  "Gets all sensitive fields that should be redacted in API responses. This includes everything from the
+   sensitive-fields set, along with any custom properties from connection-properties which are marked with a
+   :type :password."
```
https://guide.clojure.style/#docstring-indentation

jeff303 replied:
Fixed in the refactor.

camsaul commented on src/metabase/models/database.clj:
```diff
@@ -154,10 +154,14 @@
+  using the given database's driver, if that driver is valid and registered. Refer to get-sensitive-fields Scaladoc
+  for full details."
+  [database]
+  (if-not (:subprotocol database)
+    driver/default-sensitive-fields ;; preserve existing behavior by returning the default sensitive fields
+    (driver/get-sensitive-fields (driver.u/database->driver database))))
```
`database` won't have a `:subprotocol`. Database isn't a JDBC spec, it's a row from the `metabase_database` table in the app DB. The way you did it below with `(driver.u/database->driver database)` is the right way to do it.

jeff303 replied:
> * Not all data warehouse DBs we support are JDBC DBs (for example, MongoDB) -- so don't try to use `:subprotocol`. Use `driver.u/database->driver` or just `Database.engine` (these do the same thing) to get the `driver`
> * I don't think this really makes sense as a multimethod, if we can't think of any situation where someone would actually need to override the behavior. I'd rather only add multimethods (and make the driver interface bigger) if we can think of an actual reason someone might possibly need to override it. Since it just derives its values from `connection-properties`, a driver author should be marking things as sensitive _there_. It could be confusing if it seemed like there were two places they needed to do it. I'd make this a utility function instead
> 
> Besides those minor code organization suggestions this looks good

All of this should be addressed now (no more multimethod, just a single utility fn).

camsaul commented on src/metabase/api/database.clj:
```diff
@@ -516,7 +516,7 @@
                 details))
             details
-            database/sensitive-fields))))
+            (database/get-sensitive-fields-for-db database)))))
```
I'd just do 

```clj
(driver/get-sensitive-fields (driver.u/database->driver database))
```

directly here, since the notion of "sensitive fields" isn't really specific to an individual database but rather to an individual driver... it doesn't make sense to write a utility function in `models.database` for what is effectively just the one-liner above

jeff303 replied:
Yeah this was mostly to dance around some weird tests that were passing in null and empty maps for DB specs.  I'll figure out another way to make them happy.

camsaul commented on src/metabase/driver.clj:
```diff
@@ -594,3 +594,31 @@
+  "Gets all sensitive fields that should be redacted in API responses. This includes everything from the
+   sensitive-fields set, along with any custom properties from connection-properties which are marked with a
+   :type :password."
+  [driver]
+  (let [all-fields      (connection-properties driver)
+        password-fields (filter #(= (get % :type) :password) all-fields)]
+    (into default-sensitive-fields (map (comp keyword :name) password-fields))))
```
Is there any foreseeable situation where a driver is going to do something different than this exact behavior? Now that I think about it a bit more, I think this might be better off as a utility function in `metabase.driver.util` or something like that. We should be very careful about introducing new methods to the driver API, we have more than enough already and we don't need to make things more complicated if we don't have a good reason for doing so.

jeff303 replied:
Fixed in refactor (now just a single utility function, not a multimethod).

camsaul reviewed (changes requested):
- Not all data warehouse DBs we support are JDBC DBs (for example, MongoDB) -- so don't try to use `:subprotocol`. Use `driver.u/database->driver` or just `Database.engine` (these do the same thing) to get the `driver`
- I don't think this really makes sense as a multimethod, if we can't think of any situation where someone would actually need to override the behavior. I'd rather only add multimethods (and make the driver interface bigger) if we can think of an actual reason someone might possibly need to override it. Since it just derives its values from `connection-properties`, a driver author should be marking things as sensitive _there_. It could be confusing if it seemed like there were two places they needed to do it. I'd make this a utility function instead

Besides those minor code organization suggestions this looks good

jeff303 committed:
Extend the definition of sensitive-fields in the Database model to include any custom password type properties

Adding new utility function to determine sensitive fields for a driver, which lives in the driver.utility namespace

Delegating all calls to what used to be the sensitive-fields var to use the new fn instead

Adding new test to confirm that custom :password connection-properties are masked

camsaul reviewed (approved):
This LGTM. The only other thing I could think of is maybe you might want to make that util fun return a set so it's consistent with the detail sensitive fields

jeff303 commented:
> This LGTM. The only other thing I could think of is maybe you might want to make that util fun return a set so it's consistent with the detail sensitive fields

I think it already does.  It's either returning `default-sensitive-fields` (a set) directly, or else `(into default-sensitive-fields ...)` (which also is a set).
